pygmt/src/grd2xyz.py

```diff
@@ -0,0 +1,110 @@
+    output_type : str
+        Determine the format the xyz data will be returned in [Default is
+        ``pandas``]:
+
+            - ``numpy`` - :class:`numpy.ndarray`
+            - ``pandas``- :class:`pandas.DataFrame`
+            - ``file`` - ASCII file (requires ``outfile``)
+    outfile : str
+        The file name for the output ASCII file.
+    {R}
+        Adding `region` will select a subsection of the grid. If this
+        subsection exceeds the boundaries of the grid, only the common region
+        will be output.
+    {V}
+
+    Returns
+    -------
+    ret : pandas.DataFrame or numpy.ndarray or None
+        Return type depends on ``outfile`` and ``output_type``:
+
+        - None if ``outfile`` is set (output will be stored in file set by
+          ``outfile``)
+        - :class:`pandas.DataFrame` or :class:`numpy.ndarray` if ``outfile`` is
+          not set (depends on ``output_type`` [Default is
+          :class:`pandas.DataFrame`])
+
+    """
+    if output_type not in ["numpy", "pandas", "file"]:
+        raise GMTInvalidInput("Must specify format as either numpy, pandas, or file.")
+    if outfile is not None and output_type != "file":
+        msg = (
+            f"Changing `output_type` of grd2xyz from '{output_type}' to 'file' "
+            "since `outfile` parameter is set. Please use `output_type='file'` "
+            "to silence this warning."
+        )
+        warnings.warn(msg, category=RuntimeWarning, stacklevel=2)
+        output_type = "file"
+    elif output_type == "file" and outfile is None:
+        raise GMTInvalidInput("""Must specify outfile for ASCII output.""")
+    if "o" not in kwargs.keys():  # Only set column names for standard output
+        # Set the default column names for the pandas dataframe header
+        dataframe_header = ["x", "y", "z"]
+        # Set the column names to match an input DataArray as the grid
+        if isinstance(grid, xr.DataArray) and output_type == "pandas":
+            # Reverse the dims because it is rows, columns ordered.
+            dataframe_header = [grid.dims[1], grid.dims[0], grid.name]
+
+    with GMTTempFile() as tmpfile:
+        with Session() as lib:
+            file_context = lib.virtualfile_from_data(check_kind="raster", data=grid)
+            with file_context as infile:
+                if outfile is None:
+                    outfile = tmpfile.name
+                arg_str = " ".join([infile, build_arg_string(kwargs), "->" + outfile])
+                lib.call_module("grd2xyz", arg_str)
+
+        # Read temporary csv output to a pandas table
+        if outfile == tmpfile.name:  # if user did not set outfile, return pd.DataFrame
+            if "o" not in kwargs.keys():
+                result = pd.read_csv(
+                    tmpfile.name, sep="\t", names=dataframe_header, comment=">"
+                )
+            else:
+                result = pd.read_csv(tmpfile.name, sep="\t", comment=">")
```

Comment:
Need a test with `o` option to ensure this line has code coverage.

Reply:
@weiji14 @seisman I'm struggling to come up with a good test for this. Should it just assert that some of the columns will be `NaN` if an argument for `o` is provided?

Reply:
When `-o` is used, only some of the `x`, `y`, `z` columns will be written to the temporary file. For this case, it's difficult to guess which columns are written unless PyGMT can parse the `-o` option. 

I'm thinking if it's simpler to disable the `-o` option for pandas output.

Reply:
> I'm thinking if it's simpler to disable the `-o` option for pandas output.

I like this. I don't know why someone wouldn't want all of the information in the output dataframe; the user can use `pandas` functions to select some of the columns.